### PR TITLE
Allow configuration by file which will allow multiple rmux pools

### PR DESCRIPTION
**Pull request by asuahsahua**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -84,7 +84,7 @@ func (myClient *Client) ParseCommand(firstLine []byte, isMultiplexing bool) (res
 		protocol.Debug("Received error from GetCommand: %s\r\n", err)
 		return false, err
 	}
-//	protocol.Debug("Received %s %s\r\n", myClient.command[0:myClient.commandLength], myClient.firstArgument)
+	protocol.Debug("Received %s %s\r\n", myClient.command[0:myClient.commandLength], myClient.firstArgument)
```
Comment (mikemfrank) [inline client.go]: Do we want to leave this uncommented?

Comment (asuahsahua) [inline client.go]: Without the dev flag this should end up doing nothing (https://github.com/forcedotcom/rmux/blob/d28bbcda5191004c03e023f875296eee1fb2d18a/protocol/prod_debug.go#L21-21), but I should disassemble the binary and check to see if the function call is optimized out or not.

Comment (asuahsahua) [inline client.go]: `go tool objdump` shows
```
client.go:87    0x8d1ce e8cd820900          CALL github.com/forcedotcom/rmux/protocol.Debug(SB)
```

So looks like we should remove this line. I wonder if there's a way to remove function calls like this unless certain build flags are enabled...

File: main/main.go [hunk 2]
```diff
@@ -21,117 +21,237 @@ import (
...
+	var err error
+
 	if *cpuProfile != "" {
 		f, err := os.Create(*cpuProfile)
-		if err == nil {
-			pprof.StartCPUProfile(f)
-			defer pprof.StopCPUProfile()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error when creating cpu profile file: %s\r\n", err)
+			os.Exit(1)
```
Comment (vpothuri) [inline main/main.go]: move terminate code to new function like terminateIfError(err) and call this function wherever needed.

File: client.go (hunk 1 above)
Comment (asuahsahua) [inline client.go]: Seems it's a known issue: https://github.com/golang/go/issues/4563

File: main/main.go [hunk 3]
```diff
@@ -21,117 +21,237 @@ import (
...
+	for i, rmuxInstance := range rmuxInstances {
+		waitGroup.Add(1)
+
+		go func(instance *rmux.RedisMultiplexer) {
+			defer waitGroup.Done()
+
+			err := instance.Start()
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error starting rmux instance %d: %s", i, err)
+				return
```
Comment (vpothuri) [inline main/main.go]: should we remove it from waitGroup on error?

Comment (asuahsahua) [inline main/main.go]: the `defer` above should take care of that

File: protocol/protocol.go [hunk 4]
```diff
@@ -37,22 +37,22 @@ var (
...
-	SELECT_COMMAND      = []byte{'s', 'e', 'l', 'e', 'c', 't'}
-	QUIT_COMMAND        = []byte{'q', 'u', 'i', 't'}
+	DEL_COMMAND         = []byte("del")
+	SUBSCRIBE_COMMAND   = []byte("subscribe")
+	UNSUBSCRIBE_COMMAND = []byte("unsubscribe")
+	PING_COMMAND        = []byte("ping")
+	INFO_COMMAND        = []byte("info")
+	SHORT_PING_COMMAND  = []byte("PING")
+	SELECT_COMMAND      = []byte("select")
+	QUIT_COMMAND        = []byte("quit")
```
Comment (vpothuri) [inline protocol/protocol.go]: I think this does conversion first to bytes before creating byte array. Isn't it good idea to leave them as byte array especially if protocol is being called repetitively.

Comment (asuahsahua) [inline protocol/protocol.go]: Comparing an objdump of the two ways of doing it:
```diff
$ diff objdump.txt objdump2.txt
16743c16743
<       mprof.go:144    0x137a0 48c70424f8f81500                MOVQ $type.*+199672(SB), 0(SP)
---
>       mprof.go:144    0x137a0 48c70424f8f81500                MOVQ $runtime.rodata+199672(SB), 0(SP)
39452c39452
<       malloc.c:219    0x2aaf7 4805ffff1300            ADDQ $type.*+70399(SB), AX
---
>       malloc.c:219    0x2aaf7 4805ffff1300            ADDQ $runtime.rodata+70399(SB), AX
```

I think from this we can conclude that there isn't a performance difference between the two ways of declaring the byte arrays.

**Comment by vpothuri** [comment]
+1

**Pull request merged**